src/jsont.ml

```diff
@@ -1041,6 +1041,69 @@ module Object = struct
     let some = Map { kind = ""; doc = ""; dom; dec; enc} in
     mem ?doc ~dec_absent:None ?enc:e ~enc_omit:Option.is_none name some map
 
+  module Syntax = struct
+    type ('o, 'a, 'fn, 'res) mmaps =
+      | Mem : ('o, 'a) Mem.map -> ('o, 'a, 'a -> 'res, 'res) mmaps
+      | And : ('o, 'a, 'fn, 'gn) mmaps * ('o, 'b, 'gn, 'res) mmaps
+              -> ('o, 'a * 'b, 'fn, 'res) mmaps
+
+    type ('o, 'a, 'res) mmaps_ =
+      Exist : ('o, 'a, _, 'res) mmaps -> ('o, 'a, 'res) mmaps_
+
+    type ('o, 'a) schema =
+      Schema : { mmaps: 'res. ('o, 'b, 'res) mmaps_; extract: 'b -> 'a }
+               -> ('o, 'a) schema
+
+    let mem ?(doc = "") ?dec_absent ?enc ?enc_omit name type' =
+      let mmap = Mem.map ~doc ?dec_absent ?enc ?enc_omit name type' in
+      Schema { mmaps = Exist (Mem mmap); extract = Fun.id }
+
+    let opt_mem ?doc ?enc name type' =
+      mem ?doc ~dec_absent:None ~enc_omit:Option.is_none name (some type')
+
+    let ( let+ ) (Schema { mmaps; extract }) f =
+      Schema { mmaps ; extract = fun x -> f (extract x) }
+
+    let ( and+ ) (Schema a) (Schema b) =
+      let mmaps =
+        let Exist b = b.mmaps in
+        let Exist a = a.mmaps in
+        Exist (And (a, b))
+      in
+      Schema { mmaps; extract = fun (x, y) -> a.extract x, b.extract y }
+
+    let rec to_fn :
+      type o a fn res. (o, a, fn, res) mmaps -> (a -> res) -> fn
+    = fun mmaps k ->
+      match mmaps with
+      | Mem _ -> k
+      | And (a, b) -> to_fn a (fun a -> to_fn b (fun b -> k (a, b)))
```

Comment:
I missed that aspect, this adds quite a few allocations for applying the constructor.

Reply:
Yeah sadly the tuple allocations are unavoidable with let syntax :/ As far as I understand,

```ocaml
let+ a = f and+ b = g and+ c = h in body ;;
```
desugars to:
```ocaml
(let+) ((and+) ((and+) f g) h) (fun ((a, b), c) -> body)
```
so we have no choice but to build that `((a, b), c)` to call the constructor